cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1087,15 +1088,24 @@ private boolean isValid(RunRecordDetail existing, ProgramRunStatus nextProgramSt
    *
    * @param limit count at most that many runs, stop if there are more.
    */
-  public int countActiveRuns(@Nullable Integer limit) throws IOException {
-    AtomicInteger count = new AtomicInteger(0);
+  public Map<ProgramRunStatus, AtomicInteger> countActiveRuns(@Nullable Integer limit) throws IOException {
+    Map<ProgramRunStatus, AtomicInteger> result = new ConcurrentHashMap<>();
     try (CloseableIterator<RunRecordDetail> iterator = queryProgramRuns(
       Range.singleton(getRunRecordNamespacePrefix(TYPE_RUN_RECORD_ACTIVE, null)),
       key -> !NamespaceId.SYSTEM.getNamespace().equals(key.getString(StoreDefinition.AppMetadataStore.NAMESPACE_FIELD)),
       null, limit != null ? limit : Integer.MAX_VALUE)) {
-      iterator.forEachRemaining(m -> count.getAndIncrement());
+      iterator.forEachRemaining(m -> {
+        // if this is a program within a workflow or a preview run, counting is skipped.
+        if (m.getSystemArgs().containsKey(ProgramOptionConstants.WORKFLOW_NAME)
```

Comment:
not familiar with this part, from the code ,looks like if it's a program within a workflow, it will NOT have WORKFLOW_NAME ? (should we add a "!" before this condition?)

Reply:
It's the other way around. If a program is within a workflow, it will have WORKFLOW_NAME set in the args. Example [here](https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java#L269).